dial.go

```diff
@@ -89,99 +89,77 @@ func (d *Dialer) Dial(ctx context.Context, raddr ma.Multiaddr, remote peer.ID) (
 		return nil, ipnet.ErrNotInPrivateNetwork
 	}
 
-	var connOut iconn.Conn
-	var errOut error
-	done := make(chan struct{})
-
-	// do it async to ensure we respect don contexteone
-	go func() {
-		defer func() {
-			select {
-			case done <- struct{}{}:
-			case <-ctx.Done():
-			}
-		}()
-
-		maconn, err := d.rawConnDial(ctx, raddr, remote)
+	var err error
+	defer func() {
 		if err != nil {
-			errOut = err
-			return
-		}
-
-		if d.Protector != nil {
-			pconn, err := d.Protector.Protect(maconn)
-			if err != nil {
-				maconn.Close()
-				errOut = err
-				return
-			}
-			maconn = pconn
-		}
-
-		if d.Wrapper != nil {
-			maconn = d.Wrapper(maconn)
-		}
-
-		cryptoProtoChoice := SecioTag
-		if !iconn.EncryptConnections || d.PrivateKey == nil {
-			cryptoProtoChoice = NoEncryptionTag
+			logdial["error"] = err.Error()
+			logdial["dial"] = "failure"
 		}
+	}()
 
-		maconn.SetDeadline(deadline)
+	maconn, err := d.rawConnDial(ctx, raddr, remote)
+	if err != nil {
+		return nil, err
+	}
 
-		err = msmux.SelectProtoOrFail(cryptoProtoChoice, maconn)
+	defer func() {
 		if err != nil {
 			maconn.Close()
-			errOut = err
-			return
-		}
-
-		maconn.SetDeadline(time.Time{})
-
-		c := newSingleConn(ctx, d.LocalPeer, remote, maconn)
-		if d.PrivateKey == nil || !iconn.EncryptConnections {
-			log.Warning("dialer %s dialing INSECURELY %s at %s!", d, remote, raddr)
-			connOut = c
-			return
 		}
+	}()
 
-		c2, err := newSecureConn(ctx, d.PrivateKey, c)
+	if d.Protector != nil {
+		maconn, err = d.Protector.Protect(maconn)
```

Comment:
I am slightly worried that we could try to make a protector that does work here (and needs to either take a context or be called from a go routine. Thoughts @Kubuxu?

Reply:
It is possible but protectors currently work lazily, so I am not sure.

Reply:
Got it. We just need to remember to always do that.